### PR TITLE
refactor: replace char_u variables and functions with char

### PR DESCRIPTION
**Pull request by dundargoc**
Work on https://github.com/neovim/neovim/issues/459

### CODE REVIEW

File: src/nvim/mbyte.c [hunk 1]
```diff
@@ -602,14 +602,14 @@ size_t mb_string2cells_len(const char_u *str, size_t size)
 /// @param[in]  p  String to convert.
 ///
 /// @return Unicode codepoint or byte value.
-int utf_ptr2char(const char_u *const p)
+int utf_ptr2char(const char *const p)
   FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT
 {
-  if (p[0] < 0x80) {  // Be quick for ASCII.
+  if ((uint8_t)p[0] < 0x80) {  // Be quick for ASCII.
```
Comment (bfredl) [inline src/nvim/mbyte.c:608]: I wonder if we want to do this or rather do something like change `p` to `p_in` in the arg name and then `uint8_t p = (uint8_t *)p_in` on the first row. Less risk for doing mistakes (clangd warnings might not catch everything, and it is nothing wrong with regarding a string as an array of unsigned chars when we want to).

Comment (dundargoc) [reply src/nvim/mbyte.c:608]: Yep, you're right. The tests are already failing and I think this should fix it. Thanks.

Also realized I didn't have [this warning](https://clang.llvm.org/extra/clang-tidy/checks/hicpp-signed-bitwise.html) enabled. This will help save some time in the future :)

**Comment by bfredl** [comment]
needs a rebase, also to check interactions with https://github.com/neovim/neovim/pull/18416

File: src/nvim/mbyte.c [hunk 2]
```diff
@@ -889,10 +891,11 @@ int utf_ptr2len_len(const char_u *p, int size)
 /// Return the number of bytes occupied by a UTF-8 character in a string
 ///
 /// This includes following composing characters.
-int utfc_ptr2len(const char_u *const p)
+int utfc_ptr2len(const char *const p_in)
   FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
 {
-  uint8_t b0 = (uint8_t)(*p);
+  uint8_t b0 = (uint8_t)(*p_in);
+  uint8_t *p = (uint8_t *)p_in;
```
Comment (zeertzjq) [inline src/nvim/mbyte.c:898]:
```suggestion
  uint8_t *p = (uint8_t *)p_in;
  uint8_t b0 = *p_in;
```

File: src/nvim/mbyte.c [hunk 3]
```diff
@@ -889,10 +891,11 @@ int utf_ptr2len_len(const char_u *p, int size)
 /// Return the number of bytes occupied by a UTF-8 character in a string
 ///
 /// This includes following composing characters.
-int utfc_ptr2len(const char_u *const p)
+int utfc_ptr2len(const char *const p_in)
   FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
 {
-  uint8_t b0 = (uint8_t)(*p);
+  uint8_t *p = (uint8_t *)p_in;
+  uint8_t b0 = *p_in;
```
Comment (bfredl) [inline src/nvim/mbyte.c:898]: this should be `*p`

File: src/nvim/charset.c [hunk 4]
```diff
@@ -424,7 +424,7 @@ char_u *str_foldcase(char_u *str, int orglen, char_u *buf, int buflen)
   int i;
   int len = orglen;
 
-#define GA_CHAR(i) ((char_u *)ga.ga_data)[i]
+#define GA_CHAR(i) ((char *)ga.ga_data)[i]
```
Comment (bfredl) [inline src/nvim/charset.c:427]: hmm, is a bit weird to have this different type than `buf[i]` (though this entire function is a convoluted mess. all these macros can be eliminated by just setting `buf = ga.ga_data` , though that would be another PR)

**Comment by bfredl** [comment]
otherwise LGTM. Squash?
